### PR TITLE
remove safe_mul (undo #383, also cf. #1052)

### LINKED ISSUE #1052: gradients through np.where when one of branches is nan, #347 except with arctan2
I have come across an issue that I think is almost identical to #347, except with arctan2. Here's a repro:

```
def test(a):
  b = np.arctan2(a,a)
  print(b)
  
  temp = np.array([0.,1.,0.])
  c = np.where(temp>0.5,0.,b)
  print(c)
  
  return np.sum(c)

aa = np.array([3.,0.,-5.])
print(test(aa))
print(grad(test)(aa))
```

### PR DESCRIPTION
**Pull request by mattjj**
This PR removes `lax._safe_mul`, introduced in #383 as a fix for #347.  See [this detailed comment](https://github.com/google/jax/issues/1052#issuecomment-514083352) for an explanation. Fixes #1052.

The question remains: how should one solve the issues that came up in #347 and #1052? Here's the code from #347, along with the results that will print after merging this PR:

```python
import jax.numpy as np
from jax import grad

def test_fail(x):
  x = np.sqrt(np.sum(x ** 2, axis=1))
  ones = np.ones_like(x)
  x = np.where(x > 0.5, x, ones)
  return np.sum(x)

x = np.array([[1, 2], [3, 4], [0, 0]], dtype=np.float32)
print(grad(test_fail)(x))
# [[0.4472136 0.8944272]
#  [0.6       0.8      ]
#  [      nan       nan]]
```

 Here's what I consider a canonical solution at this point:

```python
def test_succeed(x):
  y = np.sum(x ** 2, axis=1)
  z = np.where(y > 0.25, y, np.ones_like(y))
  return np.sum(np.sqrt(z))

print(grad(test_succeed)(x))
# [[0.4472136 0.8944272]
#  [0.6       0.8      ]
#  [0.        0.       ]]
```

We rewrote the function! In particular, we moved the `np.sqrt` downstream, after the `np.where`.

The aspirational goal of any functional automatic differentiation system is to take as input a program that evaluates a mathematical function and to produce a new program that evaluates the mathematical derivative function, _regardless of how the input program denoted its mathematical function_. That is, there may be many different ways of writing the same function, like `test_fail` and `test_succed` here, and we'd like autodiff to succeed on all of them, producing a program that denotes the true mathematical derivative function (for all possible input values).

But we sometimes fall short of that goal, and this is one such case. See [the long comment](https://github.com/google/jax/issues/1052#issuecomment-514083352) on #1052 for a detailed description of why we must fail here.

The "fix" in #383 was wrongheaded because it meant silently producing incorrect results in some cases (again, see [the comment](https://github.com/google/jax/issues/1052#issuecomment-514083352)). It's better to produce nans on some bad denotations of a mathematical function than ever to produce incorrect results.

How about the example in #1052? I recommend rewriting this:

```python
def test(a):
  b = np.arctan2(a,a)
  temp = np.array([0.,1.,0.])
  c = np.where(temp>0.5,0.,b)
  return np.sum(c)

aa = np.array([3.,0.,-5.])
print(grad(test)(aa))
```

to this:

```python
def test2(a):
  temp = np.array([0., 1., 0.])
  a = np.where(temp > 0.5, 0., a)
  b = np.arctan2(a, a)
  c = np.where(temp > 0.5, 0., b)
  return np.sum(c)

print(grad(test2)(aa))
# [ 0.000000e+00  0.000000e+00 -7.450581e-09]
```

Again with the rewriting! Here we added another `where` to the input, like the `vectorized_cond` helper function in [the long comment](https://github.com/google/jax/issues/1052#issuecomment-514083352).

### CODE REVIEW

**Comment by mattjj** [comment]
All internal tests pass!
